base/linalg/factorization.jl

```diff
@@ -64,17 +67,34 @@ function qrfact!{T}(A::AbstractMatrix{T}; pivot=false)
     end
     QR(A, τ)
 end
-qrfact{T<:BlasFloat}(A::StridedMatrix{T}; pivot=false) = qrfact!(copy(A),pivot=pivot)
-qrfact{T}(A::StridedMatrix{T}; pivot=false) = (S = typeof(one(T)/norm(one(T)));S != T ? qrfact!(convert(AbstractMatrix{S},A), pivot=pivot) : qrfact!(copy(A),pivot=pivot))
+function qrfact!{T}(A::AbstractMatrix{T}, ::Pivot)
+    warn("pivoting only implemented for Float32, Float64, Complex64 and Complex128")
+    qrtfact!(A)
+end
+
+qrfact{T<:BlasFloat}(A::StridedMatrix{T}) = qrfact!(copy(A))
+qrfact{T<:BlasFloat}(A::StridedMatrix{T}, ::Pivot) = qrfact!(copy(A),pivot)
+
+
+stagedfunction qrfact{T}(A::StridedMatrix{T})
+    S = typeof(one(T)/norm(one(T)))
+    S != T ? :(qrfact!(convert(AbstractMatrix{S},A))) : :(qrfact!(copy(A)))
+end
+stagedfunction qrfact{T}(A::StridedMatrix{T}, ::Pivot)
+    S = typeof(one(T)/norm(one(T)))
+    S != T ? :(qrfact!(convert(AbstractMatrix{S},A),pivot)) : :(qrfact!(copy(A)),pivot)
+end
```

Comment:
Since this is likely to be useful outside of Base, perhaps we could have:

```julia
copy{T}(::Type{T}, x::T) = copy(x)
copy{T}(::Type{T}, x) = convert(T, x)
```

Although the second method may be too broad, since `convert` to a new type is not guaranteed to copy everything that `copy` would.